### PR TITLE
ZIMBRA-105: Zimbra Solr schema and plugins

### PR DESCRIPTION
**Pull request by ilyusha**
This PR represents the first checkpoint in the effort to externalize ZCS search to Solr.

See http://wiki.eng.zimbra.com/index.php/Zimbra/BuildingSolrBranch for instructions on how to deploy Solr with this configuration.

Work presented here can be divided into two parts:
1. Solr configuration files
2. Java plugins for custom analyzers, query parsers, and request handlers

### solr.xml ###
This is the global Solr configuration file. Aside from classpath configuration, this file contains the parameter `transientCacheSize`, currently defaulting to 100, that defines the size of the LRU cache of Solr cores to be kept loaded in memory.

### Zimbra configset ###
The _zimbra_ configset defines configuration that all indexing cores will be created with. The most important files here are:

#### solrconfig.xml ####
This is the core-level configuration file. While this file defines many things, the following are worth pointing out:
 - a `lib` directive that adds `zimbraplugins`, `zm-common`, and their 3rd party dependencies to the core's classpath
 - `ClassicIndexSchemaFactory` enabled to support `schema.xml` schema definitions
 - `autoSoftCommit` set to trigger on every new document, to ensure that newly-committed docs are immediately available for searching.
 - `autoCommit` set to 15 second intervals
 - custom `zimbrawildcard` query parser that introduces leading wildcard support (https://bugzilla.zimbra.com/show_bug.cgi?id=14292)
 - custom `zbnr` request handler mapped to _com.zimbra.solr.ZimbraBnRHandler_ class. (This has not been tested)
 - custom `commitcount` request handler mapped to _com.zimbra.solr.CommitCountHandler_ class. This may not be necesssary.
 - `postCommit` event listener set to our custom `CommitEventListener` class (though this is not currently used, so this may be possible to remove later)

#### schema.xml ####
This file contains the schema definition. For the most part, the fields named here correspond with field names in the _LuceneFields_ class. The exception is fields generated using `<copyField>` directives, which have no corresponding entry in _LuceneFields_.

Custom field analyzers are defined here in `<fieldType>` elements.

## Java Plugins ##
Custom analyzers previously defined in the `com.zimbra.cs.index.anaysis` package in `zm-mailbox/store` have been moved to this repo, and can be built with the `ant jar` command; this creates  `solrplugins.jar` that needs to be included on the Solr classpath.
In addition to field analyzers, this also includes _WildcardQueryParser_, _CommitCountHandler_, and _ZimbraBnRHandler_ classes mentioned above.

### TODO ###
This is still a work in progress. Some known areas of improvement are:
 - Add ability to enable/disable leading wildcard queries, as enabling them roughly doubles the index size and should therefore be optional
 - Add ability to configure _transientCacheSize_
 - Tune core parameters to avoid OOM errors under heavy load
 - Re-evaluate need for  _ZimbraBnRHandler_ and _CommitCountHandler_
 - Add dynamic field definitions to the schema to allow the event logger to index arbitrary fields
 - Update `schema.xml` to allow for loading runtime plugins from the Blob Store API using `runtimelib=true`
 - Clean up `solrconfig.xml` and remove unused components

### CODE REVIEW

**Review by gordyt — APPROVED** [review]
Ilya w/out being able to do a detailed analysis of the PR I think we should go ahead and merge this WIP and continue from there.

**Review by fatkudu — APPROVED** [review]
For some of the config files, it would be good to have a feel for where they come from in the commit comment.
Thinking in particular of the commit ''zimbra' configset files "

**Review by tmclane — APPROVED** [review]

File: build.xml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	<target name="jar" depends="compile" description="Creates the jar file">
+		<!-- next line will not be required once ZimbraServer is integrated from SOLR to main -->
+<!--		<maven dir=".." goal="dependency:copy-dependencies"/> -->
+        	<antcall target="zimbra-jar">
+        	<param name="jar.file" value="${jar.file}"/>
+            <param name="implementation.title" value="Solr Plugins"/>
+        </antcall>
+    </target>
+
+	<target name="deploy-libs" description="move all required jars to the proper location">
```
Comment (grishick) [inline build.xml]: does this actually work with ivy?

Comment (ilyusha) [reply build.xml]: I should probably delete this target; we no longer deploy Solr locally so this is unnecessary.

**Pull request merged**